pkg/apis/serving/v1alpha1/kfservice_types.go

```diff
@@ -63,6 +64,15 @@ type TensorflowSpec struct {
 	Resources v1.ResourceRequirements `json:"resources,omitempty"`
 }
 
+// TensorRTSpec defines arguments for configuring TensorRT model serving.
+type TensorRTSpec struct {
+	ModelRepositoryURI string `json:"modelUri"`
```

Comment:
Any reason to deviate ModelURI -> ModelRepositoryURI?

Reply:
TensorRT requires a model repositery that has a specific format, see: https://docs.nvidia.com/deeplearning/sdk/tensorrt-inference-server-master-branch-guide/docs/model_repository.html

I was thinking we might add the ability to create a local modelrepository by downloading a model and creating the directory structure given a modelUri. This idea is tracked by #137 wdyt?

Reply:
I would prefer this approach than exposing tensorRT details to the user.

The way I see it is when we support "tensorrt", we're supporting the model format, not specifically the tensorrt model server. Our implementation detail may be to point to an existing modelrepository or manually download to a local one we create.

WDYT?

Reply:
Are you saying the URI can be a model or a repository? How would we know which one it is? 

If we restrict to modelUri only, then our abstraction is maintained and the downloader would have to download it into the repository format. This makes sense to me. 

So the options as I see them are:
1. Support modelUri - literally the URI of a model not a repository (and we do the download/cache)
2. (up for debate) also support modelRepositoryUri in a one-of semantic

Reply:
> The way I see it is when we support "tensorrt", we're supporting the model format, not specifically the tensorrt model server. Our implementation detail may be to point to an existing modelrepository or manually download to a local one we create.

Is there a specific model format for TensorRT?

Reply:
It looks to me that `modelRepository` is for multi-models use case while `modelUri` is for single model. For tensorflow serving you'd need to pass a model configuration file which contains multiple `modelUri`s like following to serve multi-models.
```
--model_config_file=s3://platform-test/model.conf
```
model.conf
```
model_config_list: {

  config: {
    name: "model1",
    base_path: "s3://platform-test/cnn_mnist_model/export",
    model_platform: "tensorflow"
  },
  config: {
     name: "model2",
     base_path: "s3://platform-test/my_mnist_model/export",
     model_platform: "tensorflow"
  }
}
```

Reply:
I'd suggest that with "TensorRT" what should be supported is the intersection of modelRepository format and scoring API, since it has underlying support for several model formats. I think that supporting the modelRepository is especially necessary because the TensorRT `plan` format is CUDA version specific so I'd expect people to almost always have an ensemble[1]. 

The reality is, when going to consume it they'll have to know which API they're using [2]. I don't think it's unreasonable to ask for some of that knowledge at deployment time.  

[1] https://docs.nvidia.com/deeplearning/sdk/tensorrt-inference-server-master-branch-guide/docs/model_repository.html#tensorrt-models
[2] https://docs.nvidia.com/deeplearning/sdk/tensorrt-inference-server-master-branch-guide/docs/protobuf_api/grpc_service.proto.html

Reply:
I don't think it's at all clear from a design perspective how we want to support TensorRT.

1. It's not clear that we want to support their data plane API (vs wrapping it)
2. It's not clear we want to support their multi-model support in the way that they've written it.
3. It's not clear that we want to expose the model repository.

Perhaps we can discuss this in a Special Topics KFServing meeting?

I know Joey from Nvidia has been interested in discussing this. He's one of the TensorRT authors.

Reply:
> Are you saying the URI can be a model or a repository? How would we know which one it is?
> 
> If we restrict to modelUri only, then our abstraction is maintained and the downloader would have to download it into the repository format. This makes sense to me.
> 
> So the options as I see them are:
> 
> 1. Support modelUri - literally the URI of a model not a repository (and we do the download/cache)
> 2. (up for debate) also support modelRepositoryUri in a one-of semantic

I think the path forward for now is #1. I don't want to go the route of #2. I think we should take on the responsibility of managing a modelRepositoryURI for the user if they provide multiple KFServices that can be colocated.

This is what @yuzisun and I have been discussing re: gpu sharing.

Just to bring us all back to the founding principles. A KFService is a "unit of model serving". That means, the interface that we provide is a 1:1 mapping between infrastructure and a users model. The implementation details of how we host those models and squeeze performance out of them is a separate issue that's hidden by KFServing.

I really want to avoid becoming a grab bag of technology wrappers. If someone wants to run TensorRT with full control and visibility into the implementation, it's easy enough to do so using a Kubernetes Deployment or Knative Service.

Reply:
@cliveseldon made an excellent suggestion to help unblock this PR (and keep it small) - keep the attribute as modelUri and disable polling the source for updates! This should help us make incremental progress.

Reply:
Sounds good!